### DISCUSSION TITLE
how to read rdkit documentation

### DISCUSSION THREAD

**Discussion by LanceKnight** [rdkit/rdkit#5074] [Q&A]
Hello, I'm a beginner and I'm wondering how to properly navigate in rdkit documentation. For example, I found online that AllChem.EmbedMolecule has a argument called useRandomCoords. I would like to know more about what this argument mean. However, when I go to the [documentation for AllChem](https://www.rdkit.org/docs/source/rdkit.Chem.AllChem.html), I couln't find a designated entry for EmbedMolecule, similar to [that in pytorch](https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html) where all arguments are explained. I've encountered several cases of having difficulty in finding the proper rdkit function documentation. I'm just wondering how I can navigate more effectively in rdkit documentation.

**Comment by greglandrum** [comment]
@LanceKnight: documentation is tricky and it's an area where we could certainly do better (note: it's an open source project and pull requests with better documentation are always welcomed!). Still, there is a fair amount of documentation available.

I would suggest starting here: https://www.rdkit.org/docs/
That includes a bunch of introductory material as well as both the Python and C++ API documentation.
If you run across something which isn't documented, please just ask!

As an aside, this actually is one of the cases where the docstring for the function does explain what the arguments are:
```
In [9]: AllChem.EmbedMolecule?
Docstring:
EmbedMolecule( (Mol)mol [, (int)maxAttempts=0 [, (int)randomSeed=-1 [, (bool)clearConfs=True [, (bool)useRandomCoords=False [, (float)boxSizeMult=2.0 [, (bool)randNegEig=True [, (int)numZeroFail=1 [, (dict)coordMap={} [, (float)forceTol=0.001 [, (bool)ignoreSmoothingFailures=False [, (bool)enforceChirality=True [, (bool)useExpTorsionAnglePrefs=True [, (bool)useBasicKnowledge=True [, (bool)printExpTorsionAngles=False [, (bool)useSmallRingTorsions=False [, (bool)useMacrocycleTorsions=False [, (int)ETversion=1]]]]]]]]]]]]]]]]]) -> int :
    Use distance geometry to obtain initial 
     coordinates for a molecule
    
     
     ARGUMENTS:
    
        - mol : the molecule of interest
        - maxAttempts : the maximum number of attempts to try embedding 
        - randomSeed : provide a seed for the random number generator 
                       so that the same coordinates can be obtained 
                       for a molecule on multiple runs. If -1, the 
                       RNG will not be seeded. 
        - clearConfs : clear all existing conformations on the molecule
        - useRandomCoords : Start the embedding from random coordinates instead of
                            using eigenvalues of the distance matrix.
        - boxSizeMult    Determines the size of the box that is used for
                         random coordinates. If this is a positive number, the 
                         side length will equal the largest element of the distance
                         matrix times boxSizeMult. If this is a negative number,
                         the side length will equal -boxSizeMult (i.e. independent
                         of the elements of the distance matrix).
        - randNegEig : If the embedding yields a negative eigenvalue, 
                       pick coordinates that correspond 
                       to this component at random 
        - numZeroFail : fail embedding if we have at least this many zero eigenvalues 
        - coordMap : a dictionary mapping atom IDs->coordinates. Use this to 
                     require some atoms to have fixed coordinates in the resulting 
                     conformation.
        - forceTol : tolerance to be used during the force-field minimization with 
                     the distance geometry force field.
        - ignoreSmoothingFailures : try to embed the molecule even if triangle smoothing
                     of the bounds matrix fails.
        - enforceChirality : enforce the correct chirality if chiral centers are present.
        - useExpTorsionAnglePrefs : impose experimental torsion angle preferences
        - useBasicKnowledge : impose basic knowledge such as flat rings
        - printExpTorsionAngles : print the output from the experimental torsion angles
    
     RETURNS:
    
        ID of the new conformation added to the molecule 
    
    

    C++ signature :
        int EmbedMolecule(RDKit::ROMol {lvalue} [,unsigned int=0 [,int=-1 [,bool=True [,bool=False [,double=2.0 [,bool=True [,unsigned int=1 [,boost::python::dict {lvalue}={} [,double=0.001 [,bool=False [,bool=True [,bool=True [,bool=True [,bool=False [,bool=False [,bool=False [,unsigned int=1]]]]]]]]]]]]]]]]])
```

**Reply by LanceKnight** [reply]
Thanks @greglandrum for the detailed answer!
However, when I try to search for EmbedMolecule, I found it under the module rdDistGeom in[ this link](https://www.rdkit.org/docs/source/rdkit.Chem.rdDistGeom.html) instead of the AllChem you mentioned. Is it a version thing, or somehow rdKistGeom is related to AllChem, and how should I learn about those relationship between modules?

Btw, what does "eigenvalues of the distance matrix" here mean? I know the eignvalues for the matrix, but I'm not sure how it is used for calculating the conformation.

**Reply by greglandrum** [reply]
> However, when I try to search for EmbedMolecule, I found it under the module rdDistGeom in[ this link](https://www.rdkit.org/docs/source/rdkit.Chem.rdDistGeom.html) instead of the AllChem you mentioned. Is it a version thing, or somehow rdKistGeom is related to AllChem, and how should I learn about those relationship between modules?

AllChem is a convenience module which imports a bunch of other things from other modules. The API documentaiton always refers to the original module.

**Reply by greglandrum** [reply]
> Btw, what does "eigenvalues of the distance matrix" here mean? I know the eignvalues for the matrix, but I'm not sure how it is used for calculating the conformation.

The basic details of the conformer generation algorithm are in the documentation along with references to where you can get more information: https://www.rdkit.org/docs/GettingStartedInPython.html#working-with-3d-molecules

**Reply by LanceKnight** [reply]
Fantastic! Thank you so much!
